src/Microsoft.CodeAnalysis.ResxSourceGenerator/Microsoft.CodeAnalysis.ResxSourceGenerator/AbstractResxGenerator.cs

```diff
@@ -22,166 +23,164 @@
+                    }
+                });
+        }
 
-            var impl = new Impl(
-                this,
-                language: context.Compilation.Language,
-                resourceFile: resourceFile,
-                resourceName: string.Join(".", rootNamespace, resourceName),
-                resourceClassName: null,
-                omitGetResourceString: omitGetResourceString,
-                asConstants: asConstants,
-                includeDefaultValues: includeDefaultValues,
-                emitFormatMethods: emitFormatMethods);
-            if (impl.Execute(context))
+        private static void ProcessResourceFile(SourceProductionContext context, ResourceInformation resourceFile)
+        {
+            var impl = new Impl(resourceFile);
+            if (impl.Execute(context.CancellationToken))
             {
                 context.AddSource(impl.OutputTextHintName!, impl.OutputText);
             }
         }
 
-        private class Impl
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="AssemblyName"></param>
+        /// <param name="CodeLanguage">Language of source file to generate. Supported languages: CSharp, VisualBasic.</param>
+        /// <param name="SupportsNullable"></param>
+        private sealed record CompilationInformation(
+            string? AssemblyName,
+            string CodeLanguage,
+            bool SupportsNullable,
+            bool HasAggressiveInlining,
+            bool HasNotNullIfNotNull);
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="CompilationInformation">Information about the compilation.</param>
+        /// <param name="ResourceFile">Resources (resx) file.</param>
+        /// <param name="ResourceName">Name of the embedded resources to generate accessor class for.</param>
+        /// <param name="ResourceClassName">Optionally, a <c>namespace.type</c> name for the generated Resources accessor class. Defaults to <see cref="ResourceName"/> if unspecified.</param>
+        /// <param name="OmitGetResourceString">If set to <see langword="true"/>, the <c>GetResourceString</c> method is not included in the generated class and must be specified in a separate source file.</param>
+        /// <param name="AsConstants">If set to <see langword="true"/>, emits constant key strings instead of properties that retrieve values.</param>
+        /// <param name="IncludeDefaultValues">If set to <see langword="true"/>, calls to <c>GetResourceString</c> receive a default resource string value.</param>
+        /// <param name="EmitFormatMethods">If set to <see langword="true"/>, the generated code will include <c>.FormatXYZ(...)</c> methods.</param>
+        private sealed record ResourceInformation(
+            CompilationInformation CompilationInformation,
+            AdditionalText ResourceFile,
```

Comment:
Does `AdditionalText` have proper equality semantics?

Reply:
https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/AdditionalTextFile.cs

https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/DiagnosticAnalyzer/AdditionalText.cs

I don't see how that would work, and the fact that there is `AdditionalTextsProvider` makes me more confused now 😄

Reply:
We have https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs but it's not used at all in product code, it's only tested, for which I opened an issue https://github.com/dotnet/roslyn/issues/63915

Reply:
It probably relies on reference equality, where the AdditionalText instance is reused by the next compilation.

Reply:
I wasn't able to find any logic about reusing `AdditionalText`s, but as long as this is the case it should be okay. Otherwise, this needs attention from the compiler team.

Reply:
@sharwell Thinking again about this:

- We already have `IncrementalGeneratorInitializationContext.CompilationProvider`, but this doesn't mean compilation is equatable, and doesn't mean the model should contain compilation objects.
- The same might go with `AdditionalText`. It might not be safe to have it in the model.

(If I understand correctly, non-equatable types are not harmful in the very first initial step, as this step is always run)

I think we need to be more confident whether `AdditionalText`s are re-used. Could that be unit-tested via `WithTrackingName` and `IncrementalStepRunReason`? I haven't done such testing before, but I think it should be possible.

Reply:
➡️ I discussed internally, and it's the _intent_ that these values be preserved by identity if they haven't changed content. Some of the upcoming work on incremental testing and visualization will improve the validation of this.